internal/migration/migration.go

```diff
@@ -0,0 +1,478 @@
+	}
+
+	logger.Infow("buffer size automatically inferred from existing DB", "bufferSize", bufferSize)
+
+	return bufferSize, nil
+}
+
+//nolint:lll // This function has the right amount of
+func shouldMigrate(ctx context.Context, logger log.Logger, sourceBeaconPath, beaconName string, destination chain.StorageType, pgDSN string) error {
+	//nolint:exhaustive // We want to explicitly ignore the chain.MemDB backend since there's nothing to migrate there.
+	switch destination {
+	case chain.BoltDB:
+		return shouldMigrateBolt(sourceBeaconPath)
+	case chain.PostgreSQL:
+		return shouldMigratePostgres(ctx, logger, beaconName, pgDSN)
+	default:
+		return fmt.Errorf("unknown destination type %q for migration package", destination)
+	}
+}
+
+func shouldMigrateBolt(sourceBeaconPath string) error {
+	existingDB, err := bbolt.Open(sourceBeaconPath, ownerOnly, nil)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = existingDB.Close()
+	}()
+
+	return existingDB.View(func(tx *bbolt.Tx) error {
+		bucket := tx.Bucket(bucketName)
+		_, value := bucket.Cursor().First()
+		b := beacon{}
+		err := json.Unmarshal(value, &b)
+		if err != nil {
+			return ErrMigrationNotNeeded
+		}
+
+		return nil
+	})
+}
+
+func pgConn(ctx context.Context, logger log.Logger, beaconName, pgDSN string) (*pgdb.Store, func(), error) {
+	pgConf, err := database.ConfigFromDSN(pgDSN)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	conn, err := database.Open(ctx, pgConf)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if err := schema.Migrate(ctx, conn); err != nil {
+		err := fmt.Errorf("migrating error: %w", err)
+		return nil, nil, err
+	}
+
+	store, err := pgdb.NewStore(ctx, logger, conn, beaconName)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	cancel := func() {
+		_ = store.Close(ctx)
+		_ = conn.Close()
+	}
+
+	return store, cancel, err
+}
+
+func shouldMigratePostgres(ctx context.Context, logger log.Logger, beaconName, pgDSN string) error {
+	store, cancel, err := pgConn(ctx, logger, beaconName, pgDSN)
+	if err != nil {
+		return err
+	}
+	defer cancel()
+
+	storeLen, err := store.Len(ctx)
+	if err != nil {
+		return err
+	}
+
+	if storeLen != 0 {
+		logger.Warnw("Postgres contains beacon rounds. Skipping migration", "rounds count", storeLen)
+		return ErrMigrationNotNeeded
+	}
+
+	return nil
+}
+
+func (m *migrator) doMigrate(ctx context.Context) error {
+	defer func() {
+		finishedIn := time.Since(m.startedAt).String()
+		m.logger.Infow(
+			"Finished processing beacons",
+			"beaconName", m.beaconName,
+			"finishedIn", finishedIn,
+		)
+	}()
+
+	//nolint:exhaustive // We want to explicitly ignore the chain.MemDB backend since there's nothing to migrate there.
+	switch m.destination {
+	case chain.BoltDB:
+		go m.migrateBolt()
+	case chain.PostgreSQL:
+		go m.migratePostgres(ctx)
+	}
+
+	// Spin up the reader
+	go m.reader()
+
+	return m.wait()
+}
+
+//nolint:funlen // This function has the right length
+func (m *migrator) migratePostgres(ctx context.Context) {
```

Comment:
This function has 6 defer calls :(

Let's re-exam what is going on here. Maybe we can put all this logic (at least the first 4 defer's) in a function.

Reply:
Done.